network/p2p/protocol_state_provider.go

```diff
@@ -0,0 +1,129 @@
+package p2p
```

Comment:
I think our friends working on epochs would <3 a heads up (off-PR) on the overlap between this and the epoch_transition_test @smnzhu